File: week3/java_week3/test/QuickSort.java
public class QuickSort {
    private int partition(int[] array, int start, int end) {
        int mid = start + (end - start) / 2;
        int pivot = array[mid];
        int pivotIndex = start;

        array[mid] = array[end];
        array[end] = pivot;

        for ( ; array[pivotIndex] < array[end]; pivotIndex++ );

        for ( int index = pivotIndex + 1; index < end; index++ ) {
            if ( array[index] < array[end] ) {
                pivot = array[index];

                array[index] = array[pivotIndex];
                array[pivotIndex] = pivot;
                pivotIndex += 1;
            }
        }

        pivot = array[end];
        array[end] = array[pivotIndex];
        array[pivotIndex] = pivot;

        return pivotIndex;
    }

    public void quickSort(int array[], int start, int end) {
        int index;

        if ( start >= end ) {
            return;
        }
        index = partition(array, start, end);
        quickSort(array, start, index-1);
        quickSort(array, index+1, end);        
    }
}

File: week3/examples_week3/quicksort.c
#include <stdio.h>

void print(int arr[], int size) {
    int limit = size - 1;

    for ( int i = 0; i < limit; i++ ) {
        printf("%d ", arr[i]);
    }
    printf("%d\n", arr[limit]);
}

int partition(int array[], int lo, int hi)
{
    int mid = (hi + lo) / 2;

    int pivot = array[mid];
    array[mid] = array[hi];
    array[hi] = pivot;

    int pivotIndex = lo;

    for( ; array[pivotIndex] < array[hi]; pivotIndex++);
    for ( int iterationIndex = pivotIndex + 1; iterationIndex < hi; iterationIndex++)
    {
        if (array[iterationIndex] < array[hi])
        {
            pivot = array[iterationIndex];
            array[iterationIndex] = array[pivotIndex];
            array[pivotIndex] = pivot;

            pivotIndex += 1;
        }
    }

    pivot = array[pivotIndex];
    array[pivotIndex] = array[hi];
    array[hi] = pivot;

    return pivotIndex;
}



void quickSort(int array[], int lo, int hi) {
    // partition
    int pivotIndex;

    if ( lo >= hi ) {
        return;
    }

    pivotIndex = partition(array, lo, hi);

    quickSort(array, lo, pivotIndex - 1);
    quickSort(array, pivotIndex+1, hi);
}

int main() {
    int arr[] = {12, 34, 2, 78, 1, -12, 65};

    print(arr, 7);

    quickSort()

    return 0;
}
